solutions/USACO feb17-circlecross.mdx

```diff
@@ -2,223 +2,80 @@
-str ts(str s) { return s; }
-str ts(bool b) { 
-	#ifdef LOCAL
-		return b ? "true" : "false"; 
-	#else 
-		return ts((int)b);
-	#endif
-}
-tcT> str ts(complex<T> c) { 
-	stringstream ss; ss << c; return ss.str(); }
-str ts(V<bool> v) {
-	str res = "{"; F0R(i,sz(v)) res += char('0'+v[i]);
-	res += "}"; return res; }
-template<size_t SZ> str ts(bitset<SZ> b) {
-	str res = ""; F0R(i,SZ) res += char('0'+b[i]);
-	return res; }
-tcTU> str ts(pair<T,U> p);
-tcT> str ts(T v) { // containers with begin(), end()
-	#ifdef LOCAL
-		bool fst = 1; str res = "{";
-		for (const auto& x: v) {
-			if (!fst) res += ", ";
-			fst = 0; res += ts(x);
-		}
-		res += "}"; return res;
-	#else
-		bool fst = 1; str res = "";
-		for (const auto& x: v) {
-			if (!fst) res += " ";
-			fst = 0; res += ts(x);
-		}
-		return res;
-
-	#endif
-}
-tcTU> str ts(pair<T,U> p) {
-	#ifdef LOCAL
-		return "("+ts(p.f)+", "+ts(p.s)+")"; 
-	#else
-		return ts(p.f)+" "+ts(p.s);
-	#endif
-}
 
-// OUTPUT
-tcT> void pr(T x) { cout << ts(x); }
-tcTUU> void pr(const T& t, const U&... u) { 
-	pr(t); pr(u...); }
-void ps() { pr("\n"); } // print w/ spaces
-tcTUU> void ps(const T& t, const U&... u) { 
-	pr(t); if (sizeof...(u)) pr(" "); ps(u...); }
-
-// DEBUG
-void DBG() { cerr << "]" << endl; }
-tcTUU> void DBG(const T& t, const U&... u) {
-	cerr << ts(t); if (sizeof...(u)) cerr << ", ";
-	DBG(u...); }
-#ifdef LOCAL // compile with -DLOCAL, chk -> fake assert
-	#define dbg(...) cerr << "Line(" << __LINE__ << ") -> [" << #__VA_ARGS__ << "]: [", DBG(__VA_ARGS__)
-	#define chk(...) if (!(__VA_ARGS__)) cerr << "Line(" << __LINE__ << ") -> function(" \
-		 << __FUNCTION__  << ") -> CHK FAILED: (" << #__VA_ARGS__ << ")" << "\n", exit(0);
-#else
-	#define dbg(...) 0
-	#define chk(...) 0
-#endif
-
-void setPrec() { cout << fixed << setprecision(15); }
-void unsyncIO() { cin.tie(0)->sync_with_stdio(0); }
-// FILE I/O
-void setIn(str s) { freopen(s.c_str(),"r",stdin); }
-void setOut(str s) { freopen(s.c_str(),"w",stdout); }
-void setIO(str s = "") {
-	unsyncIO(); setPrec();
-	// cin.exceptions(cin.failbit); 
-	// throws exception when do smth illegal
-	// ex. try to read letter into int
-	if (sz(s)) setIn(s+".in"), setOut(s+".out"); // for USACO
+int main() {
+	ios_base::sync_with_stdio(0); cin.tie(0);
+	freopen("circlecross.in","r",stdin);
+	freopen("circlecross.out","w",stdout);
+	string s; cin >> s;
+	int ans = 0;
+	for (int a = 0; a < s.size(); ++a) 
+		for (int b = a+1; b < s.size(); ++b) 
+			for (int c = b+1; c < s.size(); ++c) 
+				for (int d = c+1; d < s.size(); ++d) 
+					ans += s[a] == s[c] && s[b] == s[d];
+	cout << ans;
+	return 0;
 }
+```
+
+Optimized $\mathcal{O}(N^2)$ solution
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
 
 int main() {
-	setIO("circlecross");
-	str s; re(s);
+	ios_base::sync_with_stdio(0); cin.tie(0);
+	freopen("circlecross.in","r",stdin);
+	freopen("circlecross.out","w",stdout);
+	string s; cin >> s;
+	int enter[26], exit[26];
```

Comment:
Just want to double check: is it ok (proper code etiquette) to use `exit` as a variable name (it's highlighted as if it were a keyword)?

@bqi343

Reply:
I would say no. Also, why use `memset` instead of `fill` or `fill_n`?

Reply:
let me check it real quick

Reply:
seems that there is one
https://www.geeksforgeeks.org/exit-vs-_exit-c-cpp/

Reply:
`memset` is slightly faster than `fill` iirc

Reply:
From one of the comments here it looks like `fill` compiles to `memset` with `-O2` when possible, so I don't see why you wouldn't use `fill`.

(Plus, when is zero-initializing an array going to be the main bottleneck in your program :eyes:)

https://lemire.me/blog/2020/01/20/filling-large-arrays-with-zeroes-quickly-in-c/

Reply:
We can call them `start` and `end` then.

Reply:
whoops then

```cpp
int main() {
	int end[1];
	vector<int> v{5};
	cout << end(v)[-1];
}
```

doesn't work :/

Reply:
> We can call them `start` and `end` then.

I would change the first letter of them to capital letters rather than `start` and `end` as it might be confusing to read.

Reply:
> From one of the comments here it looks like `fill` compiles to `memset` with `-O2` when possible, so I don't see why you wouldn't use `fill`.
> 
> https://lemire.me/blog/2020/01/20/filling-large-arrays-with-zeroes-quickly-in-c/
> 
> (Plus, when is zero-initializing an array going to be the main bottleneck in your program 👀)

Or we can use this as an alternative way of declaring variables: int Enter[26]{-1}, Exit[26]{-1};

Reply:
> Or we can use this as an alternative way of declaring variables: int Enter[26]{-1}, Exit[26]{-1};

Wait it don't work with -1 initialization

Reply:
So I think we shouldn't remove the `memset` function

Reply:
> whoops then
> 
> ```c++
> int main() {
> 	int end[1];
> 	vector<int> v{5};
> 	cout << end(v)[-1];
> }
> ```
> 
> doesn't work :/

ugh too many keywords!!




> I would change the first letter of them to capital letters rather than using `start` and `end` as the name of these variables since it might be confusing to read.

yea do this